### PR TITLE
Use export handler for querying by default (if possible)

### PR DESCRIPTION
**Pull request by kiranchitturi**
* Automatically use export handler if the query fields and sort field have docValues enabled

### CODE REVIEW

**Comment by kiranchitturi** [comment]
Initial draft. Still need to add tests

**Comment by kiranchitturi** [comment]
This is ready for review. I have added tests and verified manually that this works.

Summary:

If `use_export_handler` or `use_cursor_marks` options are not specified explicitly, then the method `SolrRelation.isStreamingPossible` will determine if the export handler should be used to retrieve documents from Solr based on the criteria listed below:
  * All the fields that are being fetched should have `docValues` enabled
  * If sort field is specified, it should have `docValues` enabled
  * If sort field is not specified, then the first field within the `querySchema` will be used for sorting

I have fixed another issue where `SolrStreamIterator` fails when there are multiple fq's in `SolrQuery`. I have added a fix to flatten multiple fq's in to a single fq inside `SolrStreamIterator`. 


Example:

### Index data 

```scala
val csvFileLocation="/Users/kiran/Git/spark-solr/src/test/resources/test-data/nyc_yellow_taxi_sample_1k.csv"
var csvDF = sqlContext.read.format("com.databricks.spark.csv")
  .option("header", "true")
  .option("inferSchema", "true")
  .load(csvFileLocation)

val options = Map(
 "zkhost" -> "localhost:9983",
  "collection" -> "techproducts",
  "gen_uniq_key" -> "true" // Generate unique key if the 'id' field does not exist
)

csvDF.write.format("solr").options(options).mode(org.apache.spark.sql.SaveMode.Overwrite).save
```

### Query data

```scala
val options = Map(
 "zkhost" -> "localhost:9983",
  "collection" -> "techproducts")
val df = sqlContext.read.format("solr").options(options).load
df.registerTempTable("trips")

sqlContext.sql("SELECT avg(tip_amount), avg(fare_amount) FROM trips").show()
```

Logs from the query

Important snippets:

```
16/05/03 16:15:43 INFO SolrRelation: Added sort param 'tip_amount asc' to the query
16/05/03 16:15:44 INFO SolrRDD: Using export handler to fetch documents from Solr
16/05/03 16:15:44 INFO SolrRDD: Using export handler to fetch documents from Solr
```

```
scala> sqlContext.sql("SELECT avg(tip_amount), avg(fare_amount) FROM trips").show()
16/05/03 16:15:42 INFO ParseDriver: Parsing command: SELECT avg(tip_amount), avg(fare_amount) FROM trips
16/05/03 16:15:43 INFO ParseDriver: Parse Completed
16/05/03 16:15:43 INFO SolrRelation: Constructed SolrQuery: q=*:*&rows=1000&collection=techproducts&fl=tip_amount,fare_amount
16/05/03 16:15:43 INFO SolrRelation: Added sort param 'tip_amount asc' to the query
16/05/03 16:15:43 INFO SparkContext: Starting job: show at <console>:26
16/05/03 16:15:43 INFO DAGScheduler: Registering RDD 6 (show at <console>:26)
16/05/03 16:15:43 INFO DAGScheduler: Got job 0 (show at <console>:26) with 1 output partitions
16/05/03 16:15:43 INFO DAGScheduler: Final stage: ResultStage 1 (show at <console>:26)
16/05/03 16:15:43 INFO DAGScheduler: Parents of final stage: List(ShuffleMapStage 0)
16/05/03 16:15:43 INFO DAGScheduler: Missing parents: List(ShuffleMapStage 0)
16/05/03 16:15:43 INFO DAGScheduler: Submitting ShuffleMapStage 0 (MapPartitionsRDD[6] at show at <console>:26), which has no missing parents
16/05/03 16:15:44 INFO MemoryStore: Block broadcast_0 stored as values in memory (estimated size 10.7 KB, free 10.7 KB)
16/05/03 16:15:44 INFO MemoryStore: Block broadcast_0_piece0 stored as bytes in memory (estimated size 5.6 KB, free 16.3 KB)
16/05/03 16:15:44 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on localhost:51665 (size: 5.6 KB, free: 511.1 MB)
16/05/03 16:15:44 INFO SparkContext: Created broadcast 0 from broadcast at DAGScheduler.scala:1006
16/05/03 16:15:44 INFO DAGScheduler: Submitting 2 missing tasks from ShuffleMapStage 0 (MapPartitionsRDD[6] at show at <console>:26)
16/05/03 16:15:44 INFO TaskSchedulerImpl: Adding task set 0.0 with 2 tasks
16/05/03 16:15:44 INFO TaskSetManager: Starting task 0.0 in stage 0.0 (TID 0, localhost, partition 0,PROCESS_LOCAL, 3868 bytes)
16/05/03 16:15:44 INFO TaskSetManager: Starting task 1.0 in stage 0.0 (TID 1, localhost, partition 1,PROCESS_LOCAL, 3868 bytes)
16/05/03 16:15:44 INFO Executor: Running task 1.0 in stage 0.0 (TID 1)
16/05/03 16:15:44 INFO Executor: Running task 0.0 in stage 0.0 (TID 0)
16/05/03 16:15:44 INFO Executor: Fetching http://172.16.3.6:51664/jars/com.univocity_univocity-parsers-1.5.1.jar with timestamp 1462317318786
16/05/03 16:15:44 INFO Utils: Fetching http://172.16.3.6:51664/jars/com.univocity_univocity-parsers-1.5.1.jar to /private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/fetchFileTemp7226921251806119576.tmp
16/05/03 16:15:44 INFO Executor: Adding file:/private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/com.univocity_univocity-parsers-1.5.1.jar to class loader
16/05/03 16:15:44 INFO Executor: Fetching http://172.16.3.6:51664/jars/org.apache.commons_commons-csv-1.1.jar with timestamp 1462317318786
16/05/03 16:15:44 INFO Utils: Fetching http://172.16.3.6:51664/jars/org.apache.commons_commons-csv-1.1.jar to /private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/fetchFileTemp8439936343161157400.tmp
16/05/03 16:15:44 INFO Executor: Adding file:/private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/org.apache.commons_commons-csv-1.1.jar to class loader
16/05/03 16:15:44 INFO Executor: Fetching http://172.16.3.6:51664/jars/mysql-connector-java-5.1.35-bin.jar with timestamp 1462317318785
16/05/03 16:15:44 INFO Utils: Fetching http://172.16.3.6:51664/jars/mysql-connector-java-5.1.35-bin.jar to /private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/fetchFileTemp7238971801112346797.tmp
16/05/03 16:15:44 INFO Executor: Adding file:/private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/mysql-connector-java-5.1.35-bin.jar to class loader
16/05/03 16:15:44 INFO Executor: Fetching http://172.16.3.6:51664/jars/com.databricks_spark-csv_2.10-1.4.0.jar with timestamp 1462317318785
16/05/03 16:15:44 INFO Utils: Fetching http://172.16.3.6:51664/jars/com.databricks_spark-csv_2.10-1.4.0.jar to /private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/fetchFileTemp6323229783897142111.tmp
16/05/03 16:15:44 INFO Executor: Adding file:/private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/com.databricks_spark-csv_2.10-1.4.0.jar to class loader
16/05/03 16:15:44 INFO Executor: Fetching http://172.16.3.6:51664/jars/spark-solr-2.1.0-SNAPSHOT-shaded.jar with timestamp 1462317318783
16/05/03 16:15:44 INFO Utils: Fetching http://172.16.3.6:51664/jars/spark-solr-2.1.0-SNAPSHOT-shaded.jar to /private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/fetchFileTemp6016866811411612414.tmp
16/05/03 16:15:44 INFO Executor: Adding file:/private/var/folders/06/kxr2yc0n6c3fd9_21ns820kw0000gq/T/spark-9fc35988-2d25-4280-806e-a4e8678ea4df/userFiles-381a1f98-b8b8-4f60-aaf6-abad7ab78b53/spark-solr-2.1.0-SNAPSHOT-shaded.jar to class loader
16/05/03 16:15:44 INFO SolrRDD: Computing the partition 1 on host name localhost
16/05/03 16:15:44 INFO SolrRDD: Computing the partition 0 on host name localhost
16/05/03 16:15:44 INFO SolrRDD: Using the shard url http://172.16.3.6:8983/solr/techproducts_shard2_replica1/ for getting partition data for split: 1
16/05/03 16:15:44 INFO SolrRDD: Using the shard url http://172.16.3.6:8983/solr/techproducts_shard1_replica1/ for getting partition data for split: 0
16/05/03 16:15:44 INFO SolrRDD: Using export handler to fetch documents from Solr
16/05/03 16:15:44 INFO SolrRDD: Using export handler to fetch documents from Solr
16/05/03 16:15:44 INFO GenerateMutableProjection: Code generated in 117.533841 ms
16/05/03 16:15:44 INFO GenerateUnsafeProjection: Code generated in 12.091694 ms
16/05/03 16:15:44 INFO GenerateMutableProjection: Code generated in 18.219427 ms
16/05/03 16:15:44 INFO GenerateUnsafeRowJoiner: Code generated in 9.07861 ms
16/05/03 16:15:44 INFO GenerateUnsafeProjection: Code generated in 8.461028 ms
16/05/03 16:15:44 INFO SolrRDD: Fetched 499 rows from shard http://172.16.3.6:8983/solr/techproducts_shard2_replica1/ for partition 1
16/05/03 16:15:44 INFO SolrRDD: Fetched 500 rows from shard http://172.16.3.6:8983/solr/techproducts_shard1_replica1/ for partition 0
16/05/03 16:15:44 INFO Executor: Finished task 0.0 in stage 0.0 (TID 0). 1405 bytes result sent to driver
16/05/03 16:15:44 INFO Executor: Finished task 1.0 in stage 0.0 (TID 1). 1405 bytes result sent to driver
16/05/03 16:15:44 INFO TaskSetManager: Finished task 1.0 in stage 0.0 (TID 1) in 767 ms on localhost (1/2)
16/05/03 16:15:44 INFO TaskSetManager: Finished task 0.0 in stage 0.0 (TID 0) in 789 ms on localhost (2/2)
16/05/03 16:15:44 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool 
16/05/03 16:15:44 INFO DAGScheduler: ShuffleMapStage 0 (show at <console>:26) finished in 0.797 s
16/05/03 16:15:44 INFO DAGScheduler: looking for newly runnable stages
16/05/03 16:15:44 INFO DAGScheduler: running: Set()
16/05/03 16:15:44 INFO DAGScheduler: waiting: Set(ResultStage 1)
16/05/03 16:15:44 INFO DAGScheduler: failed: Set()
16/05/03 16:15:44 INFO DAGScheduler: Submitting ResultStage 1 (MapPartitionsRDD[10] at show at <console>:26), which has no missing parents
16/05/03 16:15:44 INFO MemoryStore: Block broadcast_1 stored as values in memory (estimated size 12.1 KB, free 28.4 KB)
16/05/03 16:15:44 INFO MemoryStore: Block broadcast_1_piece0 stored as bytes in memory (estimated size 6.2 KB, free 34.6 KB)
16/05/03 16:15:44 INFO BlockManagerInfo: Added broadcast_1_piece0 in memory on localhost:51665 (size: 6.2 KB, free: 511.1 MB)
16/05/03 16:15:44 INFO SparkContext: Created broadcast 1 from broadcast at DAGScheduler.scala:1006
16/05/03 16:15:44 INFO DAGScheduler: Submitting 1 missing tasks from ResultStage 1 (MapPartitionsRDD[10] at show at <console>:26)
16/05/03 16:15:44 INFO TaskSchedulerImpl: Adding task set 1.0 with 1 tasks
16/05/03 16:15:44 INFO TaskSetManager: Starting task 0.0 in stage 1.0 (TID 2, localhost, partition 0,NODE_LOCAL, 2445 bytes)
16/05/03 16:15:44 INFO Executor: Running task 0.0 in stage 1.0 (TID 2)
16/05/03 16:15:44 INFO ShuffleBlockFetcherIterator: Getting 2 non-empty blocks out of 2 blocks
16/05/03 16:15:44 INFO ShuffleBlockFetcherIterator: Started 0 remote fetches in 4 ms
16/05/03 16:15:44 INFO GenerateMutableProjection: Code generated in 9.79148 ms
16/05/03 16:15:44 INFO GenerateMutableProjection: Code generated in 9.095934 ms
16/05/03 16:15:44 INFO GenerateUnsafeProjection: Code generated in 9.203319 ms
16/05/03 16:15:44 INFO GenerateSafeProjection: Code generated in 6.29179 ms
16/05/03 16:15:44 INFO Executor: Finished task 0.0 in stage 1.0 (TID 2). 1910 bytes result sent to driver
16/05/03 16:15:44 INFO TaskSetManager: Finished task 0.0 in stage 1.0 (TID 2) in 98 ms on localhost (1/1)
16/05/03 16:15:44 INFO TaskSchedulerImpl: Removed TaskSet 1.0, whose tasks have all completed, from pool 
16/05/03 16:15:44 INFO DAGScheduler: ResultStage 1 (show at <console>:26) finished in 0.099 s
16/05/03 16:15:44 INFO DAGScheduler: Job 0 finished: show at <console>:26, took 1.002092 s
+------------------+-----------------+
|               _c0|              _c1|
+------------------+-----------------+
|1.6300500500500505|12.27087087087087|
+------------------+-----------------+

```

File: src/main/scala/com/lucidworks/spark/SolrRelation.scala [hunk 1]
```diff
@@ -146,9 +146,18 @@ class SolrRelation(
 
     try {
       val querySchema = if (!fields.isEmpty) SolrRelationUtil.deriveQuerySchema(fields, baseSchema) else schema
-      val docs = solrRDD.query(query)
-      val rows = SolrRelationUtil.toRows(querySchema, docs)
-      rows
+      if (!solrRDD.exportHandler.getOrElse(false) && !conf.useCursorMarks.getOrElse(false)) {
+        // Determine whether to use Streaming API (/export handler) when 'use_export_handler' option is not set
+        val useStreamingAPI: Boolean = SolrRelation.isStreamingPossible(querySchema, baseSchema, query)
+        val docs = solrRDD.useExportHandler(useStreamingAPI).query(query)
```
Comment (thelabdude) [inline src/main/scala/com/lucidworks/spark/SolrRelation.scala]: Let's add an INFO log message here telling the user the outcome of the isStreamingPossible check

File: src/main/scala/com/lucidworks/spark/SolrRelation.scala [hunk 2]
```diff
@@ -323,5 +335,43 @@ object SolrRelation {
...
+        } else {
+          log.warn("The sort field '" + sortField + "' does not exist in the base schema")
+          return false
+        }
+      }
+    } else {
+      // If no sort query is specified, then we will use one of the fields from the querySchema
+      val sortField = querySchema.fieldNames(0)
+      query.addSort(sortField, SolrQuery.ORDER.asc)
+      log.info("Added sort param '" + query.getSortField + "' to the query")
```
Comment (thelabdude) [inline src/main/scala/com/lucidworks/spark/SolrRelation.scala]: this breaks the "functional" contract in that it's causing a side-effect on the query. Maybe move that code out of this method and only decide true or false in isStreamingPossible

**Pull request merged**